static/js/lib/redux_api.js

```diff
@@ -0,0 +1,165 @@
+// @flow
+import { createAction } from 'redux-actions';
+import type { Dispatch } from 'redux';
+import R from 'ramda';
+
+import type { Action, Dispatcher } from '../flow/reduxTypes';
+import type { Endpoint, RestState } from '../flow/restTypes';
+import { fetchJSONWithCSRF } from './api';
+import {
+  FETCH_PROCESSING,
+  FETCH_SUCCESS,
+  FETCH_FAILURE,
+} from '../actions';
+import { GET, PATCH, POST } from '../constants';
+import { couponEndpoint } from '../reducers/coupons';
+
+const actionize = R.compose(R.toUpper, R.join("_"));
+
+export const requestActionType = (...xs: string[]) => `REQUEST_${actionize(xs)}`;
+
+export const successActionType = (...xs: string[]) => `RECEIVE_${actionize(xs)}_SUCCESS`;
+
+export const failureActionType = (...xs: string[]) => `RECEIVE_${actionize(xs)}_FAILURE`;
+
+export const clearActionType = (...xs: string[]) => `CLEAR_${actionize(xs)}`;
+
+const INITIAL_STATE: RestState = {
+  loaded: false,
+  processing: false,
+};
+
+const defaultRESTPrefixes = { GET, PATCH, POST };
+
+const getPrefixForEndpoint = (verb, endpoint) => (
+  R.propOr(`${R.toLower(verb)}Prefix`, defaultRESTPrefixes[verb], endpoint)
+);
+
+export function makeFetchFunc(endpoint: Endpoint): (...args: any) => Promise<*> {
+  return (...args) => fetchJSONWithCSRF(endpoint.url, endpoint.makeOptions(...args));
+}
+
+// we can derive an action from an Endpoint object and an HTTP verb 
+// this returns an object containing the derived REST action (which
+// is an async action suitable for use with redux-thunk), and all the
+// derived action types.
+export const deriveAction = (endpoint: Endpoint, verb: string): ((...params: any) => Dispatcher<*>) => {
+  const prefix = getPrefixForEndpoint(verb, endpoint);
+
+  const requestType = requestActionType(prefix, endpoint.name);
+  const successType = requestActionType(prefix, endpoint.name);
+  const failureType = failureActionType(prefix, endpoint.name);
+  const clearType = clearActionType(prefix, endpoint.name);
+
+  const requestAction = createAction(requestType);
+  const successAction = createAction(successType);
+  const failureAction = createAction(failureType);
+
+  const fetchFunc = R.propOr(
+    `${verb}Func`,
```

Comment:
It would also be nice for cases where we want to allow quick and dirty mocking of data for development